programs/fungible-entangler/src/instructions/swap/common.rs

```diff
@@ -0,0 +1,62 @@
+use super::arg::SwapV0Args;
+use crate::error::ErrorCode;
+use crate::state::*;
+use anchor_lang::prelude::*;
+use anchor_spl::token::TokenAccount;
+
+pub struct SwapAmount {
+  pub amount: u64,
+}
+
+pub fn swap_shared_logic(
+  parent_entangler: &Account<FungibleParentEntanglerV0>,
+  child_entangler: &Account<FungibleChildEntanglerV0>,
+  base: &Account<TokenAccount>,
+  source: &Account<TokenAccount>,
+  clock: &Sysvar<Clock>,
+  args: &SwapV0Args,
+) -> Result<SwapAmount> {
+  let amount: u64;
+  let clock = clock;
+
+  require!(
+    (args.all.is_some() && args.all == Some(true)) || args.amount.is_some(),
+    ErrorCode::InvalidArgs
+  );
+
+  require!(
+    parent_entangler.go_live_unix_time < clock.unix_timestamp,
+    ErrorCode::ParentNotLiveYet
+  );
+
+  require!(
+    child_entangler.go_live_unix_time < clock.unix_timestamp,
+    ErrorCode::ChildNotLiveYet
+  );
+
+  require!(
+    parent_entangler.freeze_swap_unix_time.is_none()
+      || (parent_entangler.freeze_swap_unix_time > Some(clock.unix_timestamp)),
+    ErrorCode::ParentSwapFrozen
+  );
+
+  require!(
+    child_entangler.freeze_swap_unix_time.is_none()
+      || (child_entangler.freeze_swap_unix_time > Some(clock.unix_timestamp)),
+    ErrorCode::ChildSwapFrozen
+  );
+
+  if args.all == Some(true) {
+    amount = if source.amount > base.amount {
+      base.amount
+    } else {
+      source.amount
+    };
+  } else {
+    amount = args.amount.unwrap();
+
+    require!(base.amount >= amount, ErrorCode::TokenAccountAmountTooLow);
```

Comment:
When swapping child tokens for parent tokens, the base is set as child_storage. This requirement then requires that there be more tokens in the child_storage than the amount you want to swap. This requirement is incorrect, because you are depositing to the child_storage and withdrawing from the parent_storage. It should be checking that the amount in the parent_storage is greater than the swap amount.

This problem seems like it is the same in both directions, i.e. also when swapping from parent to child.